Implement r1cs of bulletproofs

wa5i opened the pull request:
<!--
发起一个新Pull Request的通用原则：

1. 在PR的Description中明确说明此PR的作用
2. 如果此PR和某个Issue有关联，则需要进行显式关联，例如在PR中写明：Fixes #xxxx
3. 完成下列checklist的检查
-->

##### Checklist
<!-- 基于你的PR的实际情况移除不适用的项目。其他完成的项目修改[ ]为[x]. -->
- [ ] 在 https://yuque.com/tsdoc 增加或更新了必要的文档
- [x] 增加或更新了必要的测试用例
- [ ] 对于重要修改，更新了CHANGES文件
- [x] 当前修改存在对已有API参数或返回值的改变
- [x] 当前修改存在对旧版本功能的兼容性改变（如网络协议或密码算法）

wa5i committed:
Bulletproofs update.

1. Optimizations to the Bulletproofs algorithm
2. Distinguishing between range_proof and r1cs_proof in the Bulletproofs interface
3. Implementation of a general transcript interface.

wa5i committed:
Bulletproofs update:

1. Implement r1cs of bulletproofs.
2. Optimize data structures and functions for range proof.

wa5i committed:
Bulletproofs update.

1. Unified the witness data structure and api of r1cs and range.
2. Optimize the r1cs linear combination operation api.

InfoHunter commented on crypto/zkp/bulletproofs/r1cs.c:
```diff
@@ -0,0 +1,1196 @@
+        buf += size;
+    }
+
+    RAND_seed(seed_buf, buf - seed_buf);
```
这个需要显式调用么？

wa5i replied:
> 这个需要显式调用么？

也可以不调用，调用的目的是每个 proof 的随机数种子不一样，更随机一些

wa5i committed:
Bulletproofs update.

1. Fix the issue with the 'bulletproofs' subcommand of the app.
2. Support BP_WITNESS encoding/decoding.
3. Enhance test cases for bulletproofs app.

InfoHunter commented on crypto/zkp/bulletproofs/r1cs_linear_combination.c:
```diff
@@ -0,0 +1,607 @@
+    }
+
+    return 1;
+err:
```
失败之后，lc->items里的元素谁来释放？

wa5i replied:
> 失败之后，lc->items里的元素谁来释放？

释放 lc 的函数 BP_R1CS_LINEAR_COMBINATION_free 中会释放 lc->items 及里面的元素：sk_BP_R1CS_LINEAR_COMBINATION_ITEM_pop_free(lc->items, BP_R1CS_LC_ITEM_free);

InfoHunter replied:
That is to say, the caller needs to handle this.

InfoHunter commented on crypto/zkp/bulletproofs/r1cs_linear_combination.c:
```diff
@@ -0,0 +1,607 @@
+    }
+
+    return 1;
+err:
```
失败之后，lc->items里的元素谁来释放？

wa5i replied:
> 失败之后，lc->items里的元素谁来释放？

同上

wa5i committed:
Bulletproofs update.

1. Support r1cs function in the 'bulletproofs' subcommand of the app.
2. Support BP_R1CS_PROOF encoding/decoding.
3. Enhance test cases for bulletproofs app.

wa5i committed:
Bulletproofs update.

1. Add API: BP_R1CS_LINEAR_COMBINATION_raw_mul.
2. Add r1cs range test case.

wa5i committed:
Bulletproofs update.

1. Remove useless code.

wa5i committed:
Bulletproofs update.

1. Print variable name when printing WITNESS.

wa5i commented:
文档在这里：https://www.yuque.com/tsdoc/ts/bulletproofs ，可以结合着看

InfoHunter commented:
解决冲突后可以合并
